tools/clang/test/HLSLFileCheck/hlsl/template/InstantiateDefaultArgMember.hlsl

```diff
@@ -0,0 +1,86 @@
+// AST-NEXT: BinaryOperator {{.*}} 'const T' '+'
+// AST-NEXT: DeclRefExpr {{.*}} 'const T' lvalue ParmVar {{.*}} 'a' 'const T'
+// AST-NEXT: DeclRefExpr {{.*}} 'const T' lvalue ParmVar {{.*}} 'b' 'const T'
+
+// The second AST block is the Pupper<float> instantiation. It has a full
+// instantiation including instantiation of the default template parameter. This
+// instantiation is called both with and without the default argument.
+
+// AST: ClassTemplateSpecializationDecl {{.*}} struct Pupper definition
+// AST-NEXT: TemplateArgument type 'float'
+// AST-NEXT: CXXRecordDecl {{.*}} implicit struct Pupper
+// AST-NEXT: `-CXXMethodDecl {{.*}} used test 'float (const float, const float)'
+// AST-NEXT:   |-ParmVarDecl {{.*}} used a 'const float':'const float'
+// AST-NEXT:   |-ParmVarDecl {{.*}} used b 'const float':'const float' cinit
+// AST-NEXT:   | `-ImplicitCastExpr {{.*}} 'float':'float' <IntegralToFloating>
+// AST-NEXT:   |   `-IntegerLiteral {{.*}} 'literal int' 0
+// AST-NEXT:   `-CompoundStmt
+
+// The third AST block is the Pupper<int> instantiation. This instantiation is
+// is also a full instantiation including the initializer for the `b` parameter.
+// This function is only called with the default argument used.
+
+// AST: ClassTemplateSpecializationDecl {{.*}} struct Pupper definition
+// AST-NEXT: |-TemplateArgument type 'int'
+// AST-NEXT: |-CXXRecordDecl {{.*}} implicit struct Pupper
+// AST-NEXT: `-CXXMethodDecl {{.*}} used test 'int (const int, const int)'
+// AST-NEXT:   |-ParmVarDecl {{.*}} used a 'const int':'const int'
+// AST-NEXT:   |-ParmVarDecl {{.*}} used b 'const int':'const int' cinit
+// AST-NEXT:   | `-ImplicitCastExpr {{.*}} 'int':'int' <IntegralCast>
+// AST-NEXT:   |   `-IntegerLiteral {{.*}} 'literal int' 0
+// AST-NEXT:   `-CompoundStmt
+
+// The final AST block is the Pupper<double> instantation. This instantiation
+// does not include instantiation of the default parameter. The default
+// parameter is only instantiated when used, and this instantation is only
+// called providing both arguments.
+
+// AST: ClassTemplateSpecializationDecl {{.*}} struct Pupper definition
+// AST-NEXT: |-TemplateArgument type 'double'
+// AST-NEXT: |-CXXRecordDecl {{.*}} implicit struct Pupper
+// AST-NEXT: `-CXXMethodDecl {{.*}} used test 'double (const double, const double)'
+// AST-NEXT:   |-ParmVarDecl {{.*}} used a 'const double':'const double'
+// AST-NEXT:   |-ParmVarDecl {{.*}} used b 'const double':'const double'
+// AST-NEXT:   `-CompoundStmt 
+
+float4 main(uint vertex_id : SV_VertexID) : SV_Position
+{
+  Pupper<float> PF;
+  Pupper<int> PI;
+  Pupper<double> PD;
+  return float4(PF.test(1, 2), PF.test(4), (float)PI.test(5),
+                (float)PD.test(6, 7));
+}
+
+// CHECK: call float @"\01?test@?$Pupper@{{.*}}"(%"struct.Pupper<float>"* {{.*}}, float 1.000000e+00, float 2.000000e+00)
+// CHECK: call float @"\01?test@?$Pupper@{{.*}}"(%"struct.Pupper<float>"* {{.*}}, float 4.000000e+00, float 0.000000e+00)
```

Comment:
Would this check have failed without the change?  Or would the conversion have been constant-folded during CodeGen, making the fix undetectable in the IR here?

Reply:
In DXIL, the conversion was always constant folded so this should have succeeded before the change. Since we didn't have any tests that verified this behavior previously I felt it prudent to add one. The similar SPIR-V test case did change codegen.

The AST is different for this case though, so the AST test would have failed for this test case without the change.

Reply:
Yup, thought so, sounds good.